File: dungeons-26.1.2/src/main/java/fishmod/utils/dungeon/Split.java
package fishmod.utils.dungeon;

import fishmod.utils.Constants;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import config.practical.manager.ConfigValue;

public class Split {

    public enum TimerType {
        TICK_TIME("Tick time"), DIFFRENCE("difference");

        private final String label;

        TimerType(String label) {
            this.label = label;
        }

        @Override
        public String toString() {
            return label;
        }
    }

    public static final int GREEN = 5635925;
    public static final int GRAY = 11184810;
    public static final int DARK_GRAY = 5592405;

    @ConfigValue
    public static int realTimeColorInactive = GREEN;
    @ConfigValue
    public static int realTimeColorOngoing = GREEN;
    @ConfigValue
    public static int realTimeColorComplete = GREEN;

    @ConfigValue
    public static int serverTimeColorInactive = GRAY;
    @ConfigValue
    public static int serverTimeColorOngoing = GRAY;
    @ConfigValue
    public static int serverTimeColorComplete = GRAY;

    @ConfigValue
    public static int parenthesesColorInactive = DARK_GRAY;
    @ConfigValue
    public static int parenthesesColorOngoing = DARK_GRAY;
    @ConfigValue
    public static int parenthesesColorComplete = DARK_GRAY;

    @ConfigValue
    public static TimerType timerType = TimerType.TICK_TIME;

    private final String name, startString, endString;
    private final int color;
    private final double avg;
    private int tick;
    private long startTime, endTime;
    private boolean started, ended;

    public Split(String name, String startString, String endString, int color, double avg) {
        this.name = name;
        this.startString = startString;
        this.endString = endString;
        this.color = color;
        this.avg = avg;
        this.tick = 0;
        this.ended = false;
    }

    public double getAvg() {
        return avg;
    }

    public void parseMessage(String string) {
        if (!started) {
            if (startString.equals(string)) {
                start();
            }
        } else if (!ended) {
            if (endString.equals(string)) {
                end();
            }
        }
    }

    public void tick() {
        if (started && !ended) {
            this.tick++;
        }
    }

    public void reset() {
        tick = 0;
        ended = false;
        started = false;
    }

    public void end() {
        if (ended) return;
        endTime = System.currentTimeMillis();
        started = false;
        ended = true;
    }

    public void start() {
        startTime = System.currentTimeMillis();
        started = true;
    }

    public String getName() {
        return name;
    }

    public boolean started() {
        return started;
    }

    public boolean ended() {
        return ended;
    }

    public double getTickTime() {
        return tick * Constants.TICK_DURATION;
    }

    public double getRealTime() {
        // startTime == 0 means start() was never called; guard against returning
        // epoch-time-in-seconds when printSplits() force-ends a never-started split.
        if (startTime == 0) return 0;
        double realTime;
        if (ended) {
            realTime = (endTime - startTime) / 1000.0;
        } else if (started) {
            realTime = (System.currentTimeMillis() - startTime) / 1000.0;
        } else {
            realTime = 0;
        }
        return realTime;
    }

    public MutableComponent createNameText() {
        return Component.literal(name + " ").withColor(color);
    }

    public double getTimeDiffrence() {
        return getRealTime() - getTickTime();
    }

    public MutableComponent createTimeText() {
        int realTimeColor, serverTimeColor, parenthesesColor;

        if (!started) {
            realTimeColor = realTimeColorInactive;
            serverTimeColor = serverTimeColorInactive;
            parenthesesColor = parenthesesColorInactive;
        } else if (!ended) {
            realTimeColor = realTimeColorOngoing;
            serverTimeColor = serverTimeColorOngoing;
            parenthesesColor = parenthesesColorOngoing;
        } else {
            realTimeColor = realTimeColorComplete;
            serverTimeColor = serverTimeColorComplete;
            parenthesesColor = parenthesesColorComplete;
        }

        double tickTime = getTickTime();
        double realTime = getRealTime();

        String serverTime;
        if (timerType == TimerType.DIFFRENCE) {
            double diff = realTime - tickTime;
            if (diff > 0) {
                serverTime = "+" + Constants.DECIMAL_FORMAT.format(diff) + "s";
            } else {
                serverTime = Constants.DECIMAL_FORMAT.format(diff) + "s";
            }
        } else {
            //default to tick timer
            serverTime = Constants.DECIMAL_FORMAT.format(tickTime) + "s";
        }

        String realTimeString = (realTime >= 60? (int)(realTime / 60) + "m ": "") + Constants.DECIMAL_FORMAT.format(realTime % 60) + "s";
        return Component.literal(realTimeString).withColor(realTimeColor)
                .append(Component.literal(" (").withColor(parenthesesColor)
                        .append(Component.literal(serverTime).withColor(serverTimeColor))
                        .append(Component.literal(")").withColor(parenthesesColor)
                        ));
    }

    public void drawSplit(GuiGraphicsExtractor context, Font textRenderer, int x, int y, int maxWidth) {
        Component nameText = createNameText();
        Component timerText = createTimeText();

        int timerWidth = textRenderer.width(timerText);
        context.text(textRenderer, nameText, x, y, 0xffffffff, true);
        context.text(textRenderer, timerText, x + maxWidth - timerWidth, y, 0xffffffff, true);
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/utils/dungeon/Split.kt
package fishmod.utils.dungeon

import fishmod.shaded.practicalconfig.manager.ConfigValue
import fishmod.utils.Constants
import fishmod.utils.config.values.FishSettings
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent

class Split(
    val name: String,
    private val startString: String,
    private val endString: String,
    private val color: Int,
    val avg: Double
) {

    enum class TimerType(private val label: String) {
        TICK_TIME("Tick time"), DIFFRENCE("difference");

        override fun toString(): String = label
    }

    companion object {
        const val GREEN: Int = 5635925
        const val GRAY: Int = 11184810
        const val DARK_GRAY: Int = 5592405
        const val PB_COLOR: Int = 0xFF55FF
        const val AVG_COLOR: Int = 0xFFAA00

        @ConfigValue
        @JvmField
        var realTimeColorInactive: Int = GREEN
        @ConfigValue
        @JvmField
        var realTimeColorOngoing: Int = GREEN
        @ConfigValue
        @JvmField
        var realTimeColorComplete: Int = GREEN

        @ConfigValue
        @JvmField
        var serverTimeColorInactive: Int = GRAY
        @ConfigValue
        @JvmField
        var serverTimeColorOngoing: Int = GRAY
        @ConfigValue
        @JvmField
        var serverTimeColorComplete: Int = GRAY

        @ConfigValue
        @JvmField
        var parenthesesColorInactive: Int = DARK_GRAY
        @ConfigValue
        @JvmField
        var parenthesesColorOngoing: Int = DARK_GRAY
        @ConfigValue
        @JvmField
        var parenthesesColorComplete: Int = DARK_GRAY

        @ConfigValue
        @JvmField
        var timerType: TimerType = TimerType.TICK_TIME

        private var cachedRaw: String? = null
        private var cachedNameColors: Map<String, Int> = emptyMap()

        @JvmStatic
        fun nameColors(): Map<String, Int> {
            val raw = FishSettings.splitNameColors
            if (raw != cachedRaw) {
                cachedRaw = raw
                cachedNameColors = raw.split(';').mapNotNull { e ->
                    val i = e.lastIndexOf('=')
                    if (i <= 0) null else e.substring(i + 1).toLongOrNull(16)?.let { e.substring(0, i) to it.toInt() }
                }.toMap()
            }
            return cachedNameColors
        }

        @JvmStatic
        fun setNameColor(name: String, color: Int) {
            val m = nameColors().toMutableMap()
            m[name] = color
            FishSettings.splitNameColors = m.entries.joinToString(";") { "${it.key}=${Integer.toHexString(it.value)}" }
        }

        @JvmStatic
        fun resetNameColors() {
            FishSettings.splitNameColors = ""
        }
    }

    fun nameColor(): Int = nameColors()[name] ?: color

    private var tick: Int = 0
    private var startTime: Long = 0
    private var endTime: Long = 0
    private var started: Boolean = false
    private var ended: Boolean = false

    @JvmField var paceColor: Int = 0

    fun parseMessage(string: String) {
        if (!started) {
            if (startString == string) {
                start()
            }
        } else if (!ended) {
            if (endString == string) {
                end()
            }
        }
    }

    fun tick() {
        if (started && !ended) {
            tick++
        }
    }

    fun reset() {
        paceColor = 0
        tick = 0
        ended = false
        started = false
        cachedAt = 0L
    }

    fun end() {
        if (ended) return
        endTime = System.currentTimeMillis()
        started = false
        ended = true
        cachedAt = 0L
    }

    fun start() {
        paceColor = 0
        startTime = System.currentTimeMillis()
        started = true
        ended = false
        cachedAt = 0L
    }

    fun started(): Boolean = started

    fun ended(): Boolean = ended

    fun getTickTime(): Double = tick * Constants.TICK_DURATION

    fun getRealTime(): Double {
        if (startTime == 0L) return 0.0
        return if (ended) {
            (endTime - startTime) / 1000.0
        } else if (started) {
            (System.currentTimeMillis() - startTime) / 1000.0
        } else {
            0.0
        }
    }

    fun createNameText(): MutableComponent = Component.literal("$name ").withColor(nameColor() and 0xFFFFFF)

    fun getTimeDifference(): Double = getRealTime() - getTickTime()

    fun createTimeText(): MutableComponent {
        val realTimeColor: Int
        val serverTimeColor: Int
        val parenthesesColor: Int

        if (!started) {
            realTimeColor = realTimeColorInactive
            serverTimeColor = serverTimeColorInactive
            parenthesesColor = parenthesesColorInactive
        } else if (!ended) {
            realTimeColor = realTimeColorOngoing
            serverTimeColor = serverTimeColorOngoing
            parenthesesColor = parenthesesColorOngoing
        } else {
            val pace = when (paceColor) {
                PB_COLOR -> FishSettings.splitPbColor
                AVG_COLOR -> FishSettings.splitAvgColor
                else -> 0
            }
            realTimeColor = if (pace != 0 && FishSettings.splitPbColors) pace else realTimeColorComplete
            serverTimeColor = serverTimeColorComplete
            parenthesesColor = parenthesesColorComplete
        }

        val tickTime = getTickTime()
        val realTime = getRealTime()

        val serverTime: String = if (timerType == TimerType.DIFFRENCE) {
            val diff = realTime - tickTime
            if (diff > 0) {
                "+" + Constants.DECIMAL_FORMAT.format(diff) + "s"
            } else {
                Constants.DECIMAL_FORMAT.format(diff) + "s"
            }
        } else {
            Constants.DECIMAL_FORMAT.format(tickTime) + "s"
        }

        val realTimeString = (if (realTime >= 60) (realTime / 60).toInt().toString() + "m " else "") + Constants.DECIMAL_FORMAT.format(realTime % 60) + "s"
        return Component.literal(realTimeString).withColor(realTimeColor and 0xFFFFFF)
            .append(
                Component.literal(" (").withColor(parenthesesColor and 0xFFFFFF)
                    .append(Component.literal(serverTime).withColor(serverTimeColor and 0xFFFFFF))
                    .append(Component.literal(")").withColor(parenthesesColor and 0xFFFFFF))
            )
    }

    private var cachedName: Component? = null
    private var cachedTimer: Component? = null
    private var cachedTimerWidth = 0
    private var cachedAt = 0L

    fun drawSplit(context: GuiGraphicsExtractor, textRenderer: Font, x: Int, y: Int, maxWidth: Int) {
        val now = System.currentTimeMillis()
        val refreshMs = if (started && !ended) 50L else 500L
        var nameText = cachedName
        var timerText = cachedTimer
        if (nameText == null || timerText == null || now - cachedAt >= refreshMs) {
            nameText = createNameText()
            timerText = createTimeText()
            cachedName = nameText
            cachedTimer = timerText
            cachedTimerWidth = textRenderer.width(timerText)
            cachedAt = now
        }
        context.text(textRenderer, nameText, x, y, -0x1, true)
        context.text(textRenderer, timerText, x + maxWidth - cachedTimerWidth, y, -0x1, true)
    }
}
